File: 13.DFS.BFS/八皇后.py
#
# 测试平台 https://leetcode-cn.com/problems/eight-queens-lcci/
#
class Solution(object):
    def __init__(self):
        self.row = None
        self.left = None
        self.right = None
        self.A = None
        self.ans = None
    
    def Build(self, n):
        tmp = []
        for r in range(n):
            tmp.append(False)
        return tmp
    
    def Init(self, n):
        self.row = self.Build(n)
        self.left = self.Build(n+n)
        self.right = self.Build(n+n)
        self.A = []
        for r in range(n):
            tmp = []
            for c in range(n):
                tmp.append('.')
            self.A.append(tmp)

        self.ans = []

    def Clone(self):
        tmp = []
        for r in self.A:
            rr = ""
            for c in r:
                rr = rr + c
            tmp.append(rr)
        self.ans.append(tmp)

    def dfs(self, c, C):
        """
        c: 表示当前处理的列
        C: 表示一共有多少列，当然也是一共有多少行
        """
        if c >= C:
            self.Clone()
            return

        for r in range(C):
            if self.row[r] or self.left[r+c] or self.right[r+C-c]:
                continue

            self.row[r] = True
            self.left[r+c] = True
            self.right[r+C-c] = True
            self.A[r][c] = 'Q'
            self.dfs(c+1, C)
            self.A[r][c] = '.'
            self.row[r] = False
            self.left[r+c] = False
            self.right[r+C-c] = False

    def solveNQueens(self, n):
        self.Init(n)
        self.dfs(0, n)
        return self.ans

File: 13.DFS.BFS/八皇后.java
// https://leetcode-cn.com/problems/eight-queens-lcci/
class Solution {
    private boolean[] row = null;
    private boolean[] left = null;
    private boolean[] right = null;
    // 用来存放单次的解
    private char[][] A = null;

    // 用来存放所有的解
    private List<List<String>> ans = null;

    private void Init(int n) {
        row = new boolean[n];
        left = new boolean[n+n];
        right = new boolean[n+n];

        // 生成回溯所用的矩阵
        A = new char[n][n];
        for (int r = 0; r < n; r++) {
            Arrays.fill(A[r], '.');
        }

        // 存放所有的结果
        ans = new ArrayList<>();
    }

    private void Clone() {
        List<String> tmp = new ArrayList<>();
        for (int r = 0; r < A.length; r++) {
            tmp.add(new String(A[r]));
        }
        ans.add(tmp);
    }
    /**
     * c: start columns
     * C: total columns
     */
    private void dfs(int c, int C) {
        if (c >= C) {
            Clone();
            return;
        }

        // 当前这一列，一共有n行选择
        for (int r = 0; r < C; r++) {
            if (row[r] || left[r+c] || right[r + C -c]) {
                continue;
            }
            row[r] = left[r+c] = right[r+C-c] = true;
            A[r][c] = 'Q';
            dfs(c+1,C);
            A[r][c] = '.';
            row[r] = left[r+c] = right[r+C-c] = false;
        }
    }

    public List<List<String>> solveNQueens(int n) {
        Init(n);
        dfs(0/*start_col*/, n/*total cols*/);
        return ans;
    }
}
